src/spurt/mcf/_ortools.py

```diff
@@ -148,6 +151,71 @@ def residues_to_flows(
         """
         return solve_mcf(self._dual_edges, self._dual_edge_dir, residues, cost, revcost)
 
+    def residues_to_flows_many(
+        self,
+        residues: np.ndarray,
+        cost: np.ndarray,
+        revcost: np.ndarray | None = None,
+        worker_count: int | None = None,
+    ) -> np.ndarray:
+        """Parallel version of residues_to_flows.
+
+        The worker_count is set to get_cpu_count() - 1 if not provided.
+        Treating costs as 1D arrays for now. Can consider 2D at a later time,
+        if necessary.
+        """
+        if worker_count is None:
+            worker_count = max(1, get_cpu_count() - 1)
+
+        if revcost is None:
+            revcost = cost
+
+        # Get dimensions of the problem
+        nruns, nresidues = residues.shape
+
+        if nresidues != len(self.cycles) + 1:
+            errmsg = (
+                f"Number of residues {nresidues} does not match number of"
+                f" cycles {len(self.cycles)}"
+            )
+            raise ValueError(errmsg)
+
+        # Create flows output variable
+        flows = np.zeros((nruns, len(self.edges)), dtype=np.int32)
+
+        # Function to call with worker pool
+        def uw_inputs(idxs):
+            for ii in idxs:
+                yield (
+                    ii,
+                    self._dual_edges,
+                    self._dual_edge_dir,
+                    residues[ii],
+                    cost,
+                    revcost,
+                )
+
+        # Only use multiprocessing if needed
+        if worker_count == 1:
+            for ii, res in enumerate(residues):
+                flows[ii, :] = self.residues_to_flows(res, cost, revcost=revcost)
+
+        else:
+            # Create a pool and dispatch
+            p = Pool(processes=worker_count, maxtasksperchild=1)
+            mp_tasks = p.imap_unordered(wrap_solve_mcf, uw_inputs(range(nruns)))
+            p.close()
```

Comment:
Done

Reply:
Actually the contextmanager is really slowing things down here a lot. The unit tests take way longer. Need to read up more on this but retaining the old setup here. Probably has to do with the old fork/spawn issue.

Reply:
Hmm that's weird. How much of a difference in runtime are you seeing?

It looks like the [`__exit__()`](https://github.com/python/cpython/blob/df987331d896031799f33cb0f3a7b5c5712cb4e7/Lib/multiprocessing/pool.py#L738-L739) method (which is automatically called upon exiting the `with` block) calls `terminate()` instead of `close()`. I suspect that's the source of the discrepancy, but couldn't explain why that would cause a big slowdown.

I don't really understand the details, but the note [here](https://docs.python.org/3/library/multiprocessing.html#multiprocessing.pool.Pool) says that both `close()` and `terminate()` must be called in order to safely release the resources associated with the process pool.

>**Warning:** [multiprocessing.pool](https://docs.python.org/3/library/multiprocessing.html#module-multiprocessing.pool) objects have internal resources that need to be properly managed (like any other resource) by using the pool as a context manager or by calling [close()](https://docs.python.org/3/library/multiprocessing.html#multiprocessing.pool.Pool.close) and [terminate()](https://docs.python.org/3/library/multiprocessing.html#multiprocessing.pool.Pool.terminate) manually. Failure to do this can lead to the process hanging on finalization.
>
>Note that it is **not correct** to rely on the garbage collector to destroy the pool as CPython does not assure that the finalizer of the pool will be called (see [object.__del__()](https://docs.python.org/3/reference/datamodel.html#object.__del__) for more information).

All of the examples of `multiprocessing.Pool` in the [docs](https://docs.python.org/3/library/multiprocessing.html) are using it as a context manager, so I'm sure that's the recommended way to do it. (The usage examples don't call `close()` or `terminate()` manually.)

Reply:
Usually the tests run in 30-35 secs. I let it run for 10 mins plus and it doesn't terminate :-1:

Reply:
As suspected, the fix from isce2 works on my osx machine.

https://github.com/isce-framework/isce2/blob/2402adf509632eb25857ecb2db47b187ce4013c7/components/mroipac/ampcor/DenseAmpcor.py#L315

Reply:
Wow. Uhh OK I don't really get what's going on but I guess we'll leave it as is for now and revisit if it causes problems down the road.

Reply:
This one doesn't hang
```python
            # Create a pool and dispatch
            with Pool(processes=worker_count, maxtasksperchild=1) as p:
                mp_tasks = p.imap_unordered(wrap_solve_mcf, uw_inputs(range(nruns)))

                # (Note the indentation!)
                # Gather results
                count = 0
                for res in mp_tasks:
                    flows[res[0], :] = res[1]
                    count += 1
```

same with this version

```python
            # Create a pool and dispatch
            with Pool(processes=worker_count, maxtasksperchild=1) as p:
                mp_tasks = p.imap_unordered(wrap_solve_mcf, uw_inputs(range(nruns)))
                results = list(mp_tasks)

            # Gather results
            count = 0
            for res in results:
                flows[res[0], :] = res[1]
                count += 1
```
or using `p.map(wrap_solve_mcf,`


I think the issue with the hanging version is that `imap_unordered` returns an `IMapIterator`. This yields results, which then triggers more consumption of the input iterator. But the main process continues on unless you call something to block it (like `list(mp_tasks)`)


When the pool is automatically closed and joined, the worker processes are waiting for new data, but the main task has closed and destroyed what would feed them.

(Note what the traceback spits out when you ctrl-c the hanging version:)
<details>

```python-traceback
self = <multiprocessing.pool.IMapUnorderedIterator object at 0x739cb6671190>, timeout = None

    def next(self, timeout=None):
        with self._cond:
            try:
                item = self._items.popleft()
            except IndexError:
                if self._index == self._length:
                    self._pool = None
                    raise StopIteration from None
>               self._cond.wait(timeout)
```

</details>

Reply:
:+1: Will give with option 1 of gathering results within the context manager

Reply:
Thanks @scottstanie, nice catch